Sources/NIOHTTP2/HTTP2StreamChannel.swift

```diff
@@ -178,39 +178,63 @@ final class HTTP2StreamChannel: Channel, ChannelCore {
         self._pipeline = ChannelPipeline(channel: self)
     }
 
-    internal func configure(initializer: ((Channel, HTTP2StreamID) -> EventLoopFuture<Void>)?, userPromise promise: EventLoopPromise<Channel>?){
+    internal func configure(initializer: ((Channel, HTTP2StreamID) -> EventLoopFuture<Void>)?, userPromise promise: EventLoopPromise<Channel>?) {
         assert(self.streamDataType == .frame)
         // We need to configure this channel. This involves doing four things:
         // 1. Setting our autoRead state from the parent
         // 2. Calling the initializer, if provided.
         // 3. Activating when complete.
         // 4. Catching errors if they occur.
-        self.getAutoReadFromParent().flatMap { autoRead in
-            self.autoRead = autoRead
-            // This initializer callback can only be invoked if we already have a stream ID.
-            // So we force-unwrap here.
-            return initializer?(self, self.streamID!) ?? self.eventLoop.makeSucceededFuture(())
-        }.map {
-            self.postInitializerActivate(promise: promise)
-        }.whenFailure { error in
-            self.configurationFailed(withError: error, promise: promise)
+        self.getAutoReadFromParent().whenComplete { autoReadResult in
```

Comment:
This seems like an unfortunate patch to me. It's harder to read, more error prone (two calls to `postInitializerActivate...` and `configurationFailed...` instead of one) for a 1% reduction in instructions.

- Is this a measurable improvement in actual wall clock performance?
- do we have test coverage for all branches (because there are many more now)?
- shouldn't the old code become just as fast once we start caching `ELF<Void>`?

Reply:
Also this seems to predominantly reduce allocations so if we believe this is important, we should have an alloc counter test covering this?

Reply:
Re 3: It won't be just as fast, if only because more closures have to be constructed. And we do have alloc counter tests that cover this.

Reply:
The alloc counter tests are why the unit tests failed :wink:

Reply:
Re 3: fair enough, there'll be a small penalty. But is it worth regressing on code readability? Stream creation is obviously a hot path so I can see this being justified but I still struggle with that. The perf bot should hopefully tell us how much we gain in actual performance here.

Reply:
> This seems like an unfortunate patch to me. It's harder to read, more error prone (two calls to `postInitializerActivate...` and `configurationFailed...` instead of one) for a 1% reduction in instructions.

I agree it's unfortunate; I saw it and thought it _might_ be worth doing so threw up a PR.